accepted/install-locations.md

```diff
@@ -0,0 +1,241 @@
+
+The reason for including the default global location is so that 3.0 host can find 2.* installs since those are not registered in the system. All 3.0 and higher installs should be registered.
+
+### Framework and SDK search
+All hosts use the same logic.
+
+Both framework and SDK search uses the same logic, the only difference is that frameworks are in the `share` subdirectory, while SDKs are in the `sdk` subdirectory of the install location.
+In both cases all locations listed below are searched and the complete list of all available SDKs/Frameworks is combined before making a choice which one will be used (based on version requirements).
+
+**Important:** There's a difference in search algorithm for `hostfxr` and for frameworks/SDKs. For `hostfxr` only the first available location is considered (and the highest version from that location is selected). For frameworks/SDKs **all** locations are considered and the best match for the requested version is selected from all those locations.
+
+Search locations:
+* the location of the `hostfxr` used. See the `hostfxr` search algorithm above. This means that `DOTNET_ROOT` has effect on framework and SDK search as well since the `hostfxr` search will consider it.  
+*If the `hostfxr` is found in the `DOTNET_ROOT` location (should happen pretty much always when the variable is non-empty), then that location will be part of the search locations for frameworks/SDKs.*
+* If multi-level lookup is enabled (by default it is, can be disabled via `DOTNET_MULTILEVEL_LOOKUP=0` environment variable), then the following locations are searched as well:
+  * :new: global registered install location  
+    * Windows - `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` (32-bit view)  
+    **We've committed to support this in .NET Core 3.0 (Windows)**
+    * None on Linux/macOS
+  * default global location, in the `/host/fxr/<version>` subdirectory
+    * Windows (32-bit process on 64-bit Windows) - `%ProgramFiles(x86)%\dotnet`
+    * Windows (everything else) - `%ProgramFiles%\dotnet`
+    * None on Linux/macOS
+* Else multi-level lookup is disabled, no other search locations are considered.
+
+The reason for including the default global location is so that 3.0 host can find 2.* installs since those are not registered in the system. All 3.0 and higher installs should be registered.
+
+*Note: As of .NET Core 2.2 Linux/macOS effectively doesn't support multi-level lookup as there are no additional search paths considered when it's turned on.*
+
+The multi-level lookup feature is only useful for private installs (xcopy). Installs into global locations typically don't want/need multi-level lookup. Important scenarios:
```

Comment:
Actually the current design for Windows is that it should not happen that there are two different ones. The current design (and implementation) is that the installer checks if there's something installed in the program files location. If there is, it will only install there (and ignore any custom location which may be specified). Only if nothing is in program files it could potentially install to a different place. The registry key should point to that "one" location. The host only looks at that location, it does not perform a search across multiple folders when looking for hostfxr. Frameworks and SDKs work slightly differently, if multi-level is enabled (on windows) then potentially both the registry location and the program files will be searched (if they are different, which should happen only rarely).

The proposal here is to only modify the hostfxr search, mainly targeting the apphost scenario on Linux/Mac which currently relies on the default global location (which is not right for some distros) and on `DOTNET_ROOT`. We would not enable multi-level lookup on Linux with this change (as that is the most controversial feature and since we don't support it so far, it's better to not enable it 'by accident').

Reply:
> Actually the current design for Windows is that it should not happen that there are two different ones.

I don't see how it is possible to prevent. Why can't I install one to ProgramFiles, and then install another one and add the Registry key to point to that other one? Now I have 2.

The main point I am making above is that the more places you have to look, the more useful multi-level lookup potentially becomes. The statement in the doc says that when using the global install multi-level lookup is not useful. But now that you have 2 global install locations, it makes multi-level lookup more useful.

(Full disclosure: I'm not a fan of multi-level lookup. But I do want to be fair that there are cases where the feature does have some benefits (like in the servicing case I pointed out above)).

Reply:
It is possible to get two global locations on Windows. It should not be common at all (our installers will try to prevent it if possible). You're right that if the machine ends up with both custom global location and default location, then multi-level lookup can be useful (since it will look in both for frameworks and SDKs). I changed the doc to not rule this out, instead state that typically it's not like that.